server/src/unifyfs_inode_tree.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2019, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+ /*
+  * This file is a simple, thread-safe, segment tree implementation.  The
+  * segments in the tree are non-overlapping.  Added segments overwrite the old
+  * segments in the tree.  This is used to coalesce writes before an fsync.
+  */
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <pthread.h>
+
+#include "unifyfs_inode_tree.h"
+
+#define MIN(a, b) (a < b ? a : b)
+#define MAX(a, b) (a > b ? a : b)
+
+static int unifyfs_inode_tree_compare_func(
+    struct unifyfs_inode* node1,
+    struct unifyfs_inode* node2)
+{
+    if (node1->gfid > node2->gfid) {
+        return 1;
+    } else if (node1->gfid < node2->gfid) {
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+RB_PROTOTYPE(
+    rb_inode_tree, unifyfs_inode,
+    inode_tree_entry, unifyfs_inode_tree_compare_func)
+RB_GENERATE(
+    rb_inode_tree, unifyfs_inode,
+    inode_tree_entry, unifyfs_inode_tree_compare_func)
+
+/* Returns 0 on success, positive non-zero error code otherwise */
+int unifyfs_inode_tree_init(
+    struct unifyfs_inode_tree* tree)
+{
+    int ret = 0;
+
+    if (!tree)
+        return EINVAL;
+
+    memset(tree, 0, sizeof(*tree));
+    ret = pthread_rwlock_init(&tree->rwlock, NULL);
+    RB_INIT(&tree->head);
+
+    return ret;
+}
+
+/* Remove and free all nodes in the unifyfs_inode_tree. */
+void unifyfs_inode_tree_destroy(
```

Comment:
Are we also freeing the tree structure itself here, and if so, should we destroy the rwlock?

Reply:
Yes, you are right that we need to destroy the lock. But the structure is allocated by the caller for now.